nixos/modules/programs/fish.nix

```diff
@@ -169,6 +169,46 @@ in
       end
     '';
 
+    programs.fish.interactiveShellInit = ''
+      set fish_complete_path $fish_complete_path /etc/fish/generated_completions
+    '';
+
+    environment.etc."fish/generated_completions".source =
+      let
+        generateCompletions = package: pkgs.runCommand
+          "${package.name}-fish-completions"
+          {
+            src = package;
+            buildInputs = [ pkgs.python3 pkgs.fish ];
+          }
+          ''
+            attempt () {
+              local predicate=$1
+              shift
+              for element in "$@"; do
+                if $predicate $element; then
+                  break
+                fi
+              done
+            }
+
+            try_generate () {
+              if [ -d $1 ]; then
+                python3 ${pkgs.fish}/share/fish/tools/create_manpage_completions.py --directory $out `find $1 -type f`
```

Comment:
better to use find ... | xargs python3 here if find ever exceeds our argument list.